src/cpp/ext/otel/otel_plugin.cc

```diff
@@ -79,7 +73,71 @@ absl::flat_hash_set<std::string> BaseMetrics() {
                       kServerCallSentTotalCompressedMessageSizeInstrumentName),
       std::string(grpc::OpenTelemetryPluginBuilder::
                       kServerCallRcvdTotalCompressedMessageSizeInstrumentName)};
+  grpc_core::GlobalInstrumentsRegistry::ForEach(
+      [&](const grpc_core::GlobalInstrumentsRegistry::
+              GlobalInstrumentDescriptor& descriptor) {
+        if (descriptor.enable_by_default) {
+          base_metrics.emplace(descriptor.name);
+        }
+      });
+  return base_metrics;
 }
+
+class NPCMetricsKeyValueIterable
+    : public opentelemetry::common::KeyValueIterable {
+ public:
+  NPCMetricsKeyValueIterable(
+      absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> label_values,
+      absl::Span<const absl::string_view> optional_label_keys,
+      absl::Span<const absl::string_view> optional_label_values,
+      std::shared_ptr<std::set<absl::string_view>> enabled_optional_label_keys)
+      : label_keys_(label_keys),
+        label_values_(label_values),
+        optional_label_keys_(optional_label_keys),
+        optional_label_values_(optional_label_values),
+        enabled_optional_label_keys_(std::move(enabled_optional_label_keys)) {}
+
+  bool ForEachKeyValue(opentelemetry::nostd::function_ref<
+                       bool(opentelemetry::nostd::string_view,
+                            opentelemetry::common::AttributeValue)>
+                           callback) const noexcept override {
+    for (int i = 0; i < label_keys_.size(); i++) {
+      if (!callback(AbslStrViewToOpenTelemetryStrView(label_keys_[i]),
+                    AbslStrViewToOpenTelemetryStrView(label_values_[i]))) {
+        return false;
+      }
+    }
+    if (enabled_optional_label_keys_ == nullptr) {
+      return true;
+    }
+    // Note that if there is duplicated enabled keys we will send them multiple
+    // times.
+    for (int i = 0; i < optional_label_keys_.size(); i++) {
+      if (enabled_optional_label_keys_->find(optional_label_keys_[i]) ==
```

Comment:
We don't want to do string comparisons. As suggested in https://github.com/grpc/proposal/pull/421/files, let's use a bit_set to determine whether an optional label in a particular position needs to be recorded or not

Reply:
Yeah, we definitely want this.

Also, we need the same plumbing for the per-call metrics.  The only difference in that case is that the set of optional labels is dictated by both `AddOptionalLabels()` and by the plugin options.